### ISSUE TITLE
Tracking issue for RFC 2102, "Unnamed fields of struct and union type"

### ISSUE THREAD

**Issue by Centril** [rust-lang/rust#49804]
This is a tracking issue for the RFC "Unnamed fields of struct and union type" (rust-lang/rfcs#2102).

**Steps:**

- [ ] Implement the RFC (cc @rust-lang/compiler -- can anyone write up mentoring instructions?)
- [ ] Adjust documentation ([see instructions on forge][doc-guide])
- [ ] Stabilization PR ([see instructions on forge][stabilization-guide])

[stabilization-guide]: https://forge.rust-lang.org/stabilization-guide.html
[doc-guide]: https://forge.rust-lang.org/stabilization-guide.html#updating-documentation

**Unresolved questions:**

- [ ] [Possibly?](https://github.com/joshtriplett/rfcs/blob/unnamed-fields/text/0000-unnamed-fields.md#unresolved-questions) @joshtriplett please fill in this section :) // @centril

**Comment by joshtriplett** [comment]
@Centril Regarding unresolved questions: the first three paragraphs in that section are just talking about the future possibility of full unnamed types that can appear everywhere, which would be the subject of a future RFC. The last paragraph talks about other future possibilities that also need their own RFCs.

**Comment by Centril** [comment]
@joshtriplett Yeah; that was also my interpretation, so  ==> no unresolved questions?

**Comment by cindRoberta** [comment]
I think It, I hope it helps

RFC #2870

 think it would be better that way

let unamedstruct: { count: i32, price: f64, type: u8 } = { count: 500, price: 6.4, type: 1 };

or simply

let unamedstruct = { count: 500, price: 6.4, type: 1 };

nothing to do with tuples, starting with tuples having no named elements, structs yes

**Comment by joshtriplett** [comment]
@cindRoberta That's an unrelated proposal. This RFC is about unnamed fields within other types, not fully general anonymous types.

**Comment by ibraheemdev** [comment]
Is there any movement on this, or is it just waiting for someone to implement it?

**Comment by ibraheemdev** [comment]
@varkor maybe mentoring instructions would help someone get started?

**Comment by varkor** [comment]
I don't have time to write up full mentoring instructions at the moment, but if someone is interested in implementing this feature, I am happy to provide pointers for how to get started on Zulip.

- The first place I'd look is `parse_name_and_ty` in `compiler/rustc_parse/src/parser/item.rs`, where you'll want to start parsing an underscore, along with the rest of the syntax for unnamed fields as described in the RFC.
- This will entail modifying `StructField` in `compiler/rustc_ast/src/ast.rs` to contain one of two types: either a `NamedStructField` (containing most of what `StructField` currently contains) along with an `UnnamedStructField`, whose structure should be evident from the description in the RFC.
- Having done this, you should gate the new feature by creating a new `unnamed_fields` feature gate in `compiler/rustc_feature/src/active.rs`, which should be added to `INCOMPLETE_FEATURES`, and then checking for `UnnamedStructField` in `compiler/rustc_ast_passes/src/feature_gate.rs`.
- This should cover at least up to [**Parsing**](https://github.com/rust-lang/rfcs/blob/master/text/2102-unnamed-fields.md#parsing) in the RFC.

**Comment by jedel1043** [comment]
I'm interested on implementing this, but I need some mentoring to resolve some questions I have.

**Comment by jedel1043** [comment]
Ok, the parsing step has been completed. The next step is to lower the new types from the AST to the HIR. Could someone give me some recommendations on how to approach this?

**Comment by joshtriplett** [comment]
Is there still someone able to provide mentorship for implementing this?

Is there still someone available to try implementing this?

**Comment by nwn** [comment]
I would be interested to try implementing this, but would defer to the previous implementer if they are still interested.

**Comment by jedel1043** [comment]
> I would be interested to try implementing this, but would defer to the previous implementer if they are still interested.

Please do so! I'm somewhat busy nowadays, so I'd prefer leaving this work to someone who can dedicate their full attention to it :)

**Comment by frank-king** [comment]
Seems inactive for a while. I'd like to try implementing it (make sure that I won't duplicate @carbotaniuman's work if you are still actively working on it).

I may need help since I'm pretty new to rustc.

**Comment by carbotaniuman** [comment]
@frank-king I don't have time for a while, so feel free to take my code and use it. If you need any help with my changes you can ping me here or on zulip, but I'm not the most experienced with rustc in general.

**Comment by frank-king** [comment]
I think some unresolved problems are about the weak keyword `union` followed by an open delimiter `{` where `union` can also be parsed as an identifier.

I'm trying to enumerate the cases here, and please correct me or make a complement if there's something wrong or missing.

1. Cases that should pass the compilation (i.e. `union` should be treated as an identifier).
```rust
// `union` as a type identifier, empty `impl` item
impl union {}
// `union` as a type identifier, non-empty `impl` item
impl union {
     fn foo() {}
}
// `union` as a type identifier, empty `fn` body 
fn foo() -> union {}
// `union` as a type identifier, non-empty `fn` body 
fn foo() -> union {
    let _: () = ();
}
```
2. Cases that should be recovered as an anonymous union.
```rust
// recovered non-empty anonymous union, missing `impl` body
impl union { field: u32 }
// recovered non-empty anonymous union, empty `impl` body
impl union { field: u32 } {}
// recovered non-empty anonymous union, non-empty `impl` body
impl union { field: u32 } {
    fn foo() {}
}
// recovered empty anonymous union, empty `impl` body
impl union{} {}
// recovered empty anonymous union, non-empty `impl` body
impl union{} {
    fn foo() {}
}
// recovered non-empty anonymous union, missing `fn` body
fn foo() -> union { field: u32 }
// recovered non-empty anonymous union, empty `fn` body
fn foo() -> union { field: u32 } {}
// recovered empty anonymous union, empty `fn` body
fn foo() -> union{} {}
// recovered empty anonymous union, non-empty `fn` body
fn foo() -> union{} {
    let _: () = ();
}
```
3. Cases I'm not sure if we should recover
```rust
// syntactically correct, but probably never correct due to name conflicts
// I guess this case should be recovered so that we can report a name conflict error instead of a syntax error, is it?
impl union for union {}
// anonymous union can never be a trait
impl union {} for union {}
// but if we don't recover, the error might be confusing:
error: expected item, found keyword `for`
   | 
LL | impl union {} for union {}
   |               ^^^ expected item
```

**Comment by carbotaniuman** [comment]
Yeah, like 80% of the remaining issue with my PR was impl blocks, you can't parse anonymous unions eagerly in order not break compatibility, but you want to parse them as fallback to give nice errors. Add to that there's two places an anonymous union could or couldn't be in the impl block...

**Comment by frank-king** [comment]
I think an alternative way is to only recover the anonymous `struct`, but always treat `union` as a normal identifier when it comes to `impl` triats or types, or a function return types, leaving it a (possibly?) diagnostic issue and revisit it some time in the future.

**Comment by LcJuves** [comment]
Hope to fix it soon and implement it, got rid of it, thanks a lot!

**Comment by frank-king** [comment]
Parsing is done in #115131, and now I'm going to implement the HIR to CodeGen part.

I don't know how to assign an `AdtDef` for an anonymous adt. Could anyone give me some hints?

**Comment by Shidoengie** [comment]
i saw this and im a tad bit confused and concerned
if this means adding anonymous records and unions overall i think this a big mistake syntax wise specially on the union syntax
this would be wayyyy better
```rs
type point = {x:i32,y:132};
type intstr = i32|String;
```

**Comment by Shidoengie** [comment]
also by having this operator style union problems like `union{}` and `union{i32}` simply couldn't exist

**Comment by Shidoengie** [comment]
I haven't really touched the rust code base before however I might try to implement this

**Comment by CAD97** [comment]
For named types, I want to offer the potential alternative of using pattern destructuring as a potentially clearer alternative, e.g.

```rust
struct A { x: i32, y: i32 }
struct B {
    A { x, y }: A,
    z: i32,
}
```

(This would make `struct S { union { …` a pattern field.) I also note the RFC makes no mention of how unnamed fields of named structs should behave w.r.t. visibility of the inner struct's fields.

As an aside, a different reasonable alternative interpretation of `_: Type` as a field is to explicitly insert a `Type`’s worth of padding. It seems unnecessary to allow anonymous fields with named types. It doesn't even enable fixes to existing `#[repr(C)]` bindings to avoid compatibility breaks. IMO I think that use case is better served by a macro.

**Comment by RalfJung** [comment]
Unresolved question: what should `derive` macros do here? This applies both to the built-in ones and user-defined ones. It seems like they all need major overhaul to support types like this. And it is pretty inevitable that people *will* ask for `derive` to be supported on these types.

It's that kind of issue that makes me think that adding a new form of unnamed types to Rust (on top of closures/coroutines) is a mistake.

**Comment by fmease** [comment]
> https://github.com/rust-lang/rust/issues/49804#issuecomment-1972619108

Nominating for T-lang discussion. Context for T-lang: There's currently active compiler dev going on to implement this PR (several merged and open PRs by multiple contributors). I don't want them to continue working on it if it gets out thrown out in the end.

**Comment by Jules-Bertholet** [comment]
> The RFC also explicitly lists anonymous types as a rejected alternative, and yet the implementation that recently began for this RFC does introduce anonymous ADTs to the compiler.

I think this is the wrong implementation approach. The RFC doesn't specify the introduction of any new types. My expectation would be that this feature would be implemented by storing the field list in a tree structure instead of a flat list, for example in [`VariantDef.fields`](https://github.com/rust-lang/rust/blob/9e73597e5a83e96e223c10b3810566c74a86a0c1/compiler/rustc_middle/src/ty/mod.rs#L1173) and in `VariantData` ([ast](https://github.com/rust-lang/rust/blob/9e73597e5a83e96e223c10b3810566c74a86a0c1/compiler/rustc_ast/src/ast.rs#L2844), [hir](https://github.com/rust-lang/rust/blob/9e73597e5a83e96e223c10b3810566c74a86a0c1/compiler/rustc_hir/src/hir.rs#L2913)).

**Comment by frank-king** [comment]
Oh, previous implementation (#115732) of parsing unnamed fields as a special field kind made things very complicated and was rejected.

Was #115732 the right way to implement this feature?

**Comment by RalfJung** [comment]
If they give rise to "tree-shaped typed" then should we enforce the tree to strictly alternate between struct and union, i.e., not accept struct fields inside a struct or union fields inside a union? Those would be redundant then and give rise to questions like whether they actually change how the layout is computed.

Also, that would require pretty fundamental changes to [`FieldsShape`](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_abi/enum.FieldsShape.html), which currently encodes the assumption that a type is either struct-like or union-like.  This distinction has pretty fundamental consequences also for type validity, i.e. what the valid representations of a type are. So for instance
```rust
struct S {
  f1: bool,
  _: union { f2: u8, f3: u8 }
}
```
Should this entire type behave like a union (any byte sequence is valid), or should the first field still behave like it would inside a struct (must always be a valid `bool`)? The RFC doesn't even seem to say which fields can be safely accessed (if any) so it's not clear. If some fields can be safely accessed and others cannot, this would again be a fundamental language change, requiring the introduction of a novel kind of type to Rust.

All of this would be trivially resolved by allowing [unnamed fields with named types](https://github.com/rust-lang/rfcs/blob/master/text/2102-unnamed-fields.md#unnamed-fields-with-named-types) only. Or, you know, just naming those fields; very little code has to interact with these complicated C types directly and they can just generate names for these fields and spell them out, making the traversal of this tree-shaped structure explicit. That would also make it much easier to reason about the code, e.g. by making it clear when two fields might overlap. Given that such types are already rare in C (people seem to always use the same 2 or 3 examples when motivating this feature),  and writing C bindings is rare in Rust (in the sense that picking some random piece of Rust code, it's rare for that to be C binding code), I think a feature meant to support such a small fraction of code needs to pass a pretty high bar.

**Comment by afetisov** [comment]
> Should this entire type behave like a union (any byte sequence is valid), or should the first field still behave like it would inside a struct (must always be a valid bool)? The RFC doesn't even seem to say which fields can be safely accessed (if any) so it's not clear.

Doesn't it? Imho it's clear from the text, that the validity requirements are the same as if you had named fields with separately defined types. The only thing the RFC does is allow omitting explicit type definition and the corresponding field name in accesses:

> Given a struct s of this type, code can access s.a, s.d, and either s.b or s.c. Accesses to a and d can occur in safe code; accesses to b and c require unsafe code, and b and c overlap, requiring care to access only the field whose contents make sense at the time.

> Given a union u of this type, code can access u.a, or u.d, or both u.b and u.c. Since all of these fields can potentially overlap with others, accesses to any of them require unsafe code; however, b and c do not overlap with each other. Code can borrow u.b and u.c simultaneously, but cannot borrow any other fields at the same time.

> Such a structure defined with repr(C) will use a representation identical to the same structure with all unnamed fields transformed to equivalent named fields of a struct or union type with the same fields.

----------------------

> should we enforce the tree to strictly alternate between struct and union..?

I hope not. This feature would allow structural inheritance for structs and unions. That feature would be useful e.g. for  representation of C++-like class hierarchy, where subclasses have the fields of superclasses as their initial segments. For unions, it would allow patterns like ad-hoc subtyping of enums and hand-crafted ad-hoc sum types, like the ones OCaml uses to represent its errors.

**Comment by RalfJung** [comment]
> I hope not. This feature would allow structural inheritance for structs and unions. That feature would be useful e.g. for representation of C++-like class hierarchy, where subclasses have the fields of superclasses as their initial segments. For unions, it would allow patterns like ad-hoc subtyping of enums and hand-crafted ad-hoc sum types, like the ones OCaml uses to represent its errors.

That's outside the scope and motivation of the RFC, and the feature is likely a poor fit for this purpose. IMO that's an argument in favor of requiring strict alternation.

**Comment by afetisov** [comment]
The scope of the RFC is representation of C FFI type without introducing types and field names which don't exist in original code. Anonymous structs can appear inside structs in C, same for unions, so that is explicitly within the scope and motivation of RFC.

**Comment by simonbuchan** [comment]
FWIW I offhandedly suggested a conflicting(?) interpretation of this syntax in IRLO a short while ago:

https://internals.rust-lang.org/t/phantom-trait/20434/15?u=simonbuchan

TLDR:

```
struct Foo<T> {
  id: u32,
  _: PhantomData<T>,
}

let foo: Foo<String> { id: 123 }
```

Where the compiler can implicitly initialize some sensible set of types (`Default`? Some new `Unitary` auto trait?) without them needing to be named.

Strictly I guess this use of the syntax doesn't conflict if such types are considered to be empty types somehow?

**Comment by Amanieu** [comment]
It seems that implementation work (#121553) is blocked on the lang team nomination, so it would be nice to clarify what the exact question being asked here is.

The libc 1.0 work (https://github.com/rust-lang/libc/issues/3248) is currently blocked on this feature since we would like to make the APIs match C more closely.

This is specifically useful for APIs where the field organization can vary between platforms, where some platforms use nested unions but others use distinct fields.

**Comment by fmease** [comment]
> It seems that implementation work (#121553) is blocked on the lang team nomination, so it would be nice to clarify what the exact question being asked here is.

cc https://github.com/rust-lang/rust/pull/121270#issuecomment-2025593169

**Comment by Amanieu** [comment]
I wasn't aware that this had already been discussed since this issue is still marked as lang-nominated. I know this was a while ago, but it would be nice if @compiler-errors or someone else could summarize the position of the lang team on this feature.

**Comment by joshtriplett** [comment]
It sounds like, from the minutes, that some folks would like to see this feature designed differently than it was when it was previously accepted. It wouldn't be the first or last feature to need some design adjustments when lang design met compiler reality. Happy to help with that, so that we can find a design that meets the requirements in the original RFC and any new issues that have arisen since then.

**Comment by RalfJung** [comment]
> The libc 1.0 work (https://github.com/rust-lang/libc/issues/3248) is currently blocked on this feature since we would like to make the APIs match C more closely.
>
> This is specifically useful for APIs where the field organization can vary between platforms, where some platforms use nested unions but others use distinct fields.

Note that with the proposal as specified in the RFC, this leads to the same code being safe on some platforms and unsafe on others.

IMO some C APIs are just too terrible for us to want to support matching them in Rust. I'm not saying we shouldn't support interop with those platforms -- we always should support interop. We added untagged unions, a major non-trivial language feature, largely to support C interop. So these APIs can be translated to named structs/unions. A consistent API across platforms can be achieved via methods. That's different from what happens in C, but it is not a good idea to copy mistakes (in my eyes) from C just because we want to have APIs that look the same. "It has to match the C API" is a great recipe for being forever stuck with poor choices made decades ago. IMO a new language feature needs stronger motivation than that.

**Comment by RalfJung** [comment]
The main goal of the RFC seems to be allowing C APIs with unnamed fields to be used from Rust with essentially the same syntax. The minimal language extension I can think of that would allow this is something like this (where all names starting with `__Rust` are things that are not named in C):
```
#[repr(C)]
union __Rust_U { b: u32, c: f32 }

#[repr(C)]
struct S {
    a: u32,
    #[unnamed]
    __Rust_unnamed: U,
    d: u64,
}
```
From a Rust perspective these would be entirely normal fields and types, `derive` macros work as usual, and so on. The only thing that's special is that when doing a field lookup on type `S`, if we can't find the field directly inside `S`, we recurse into the fields of fields with the `unnamed` attribute. This desugaring happens fairly early so most of the compiler only sees the desugared form and never has to deal with unnamed fields.

This does require binding generators to come up with names for these things, but that shouldn't be too much of a problem IMO -- and seems well worth it given the amount of lang- and compiler-level trouble it avoids.

**Comment by RalfJung** [comment]
> C didn't actually have unnamed unions prior to C2011

How does that jive with the claim that unnamed fields are crucially used in some POSIX APIs and thus we need to have them in Rust to give libc the desired public API? POSIX hasn't grown much since 2011, has it?

**Comment by zackw** [comment]
> > C didn't actually have unnamed unions prior to C2011
> How does that jive with the claim that unnamed fields are crucially used in some POSIX APIs

They _are_ crucially used.  It's just that what was in the headers was (and in many cases still is) _fake_ unnamed unions, implemented with macros. I gave an example in the post you are responding to.

The _API specification_ of a C structure that may (or may not) use an an unnamed union _in the ABI_ will look like this:

```c
// The order of fields in this structure is unspecified
// Whether 'sa_handler' and 'sa_sigaction' are distinct fields is unspecified
struct sigaction {
    sigset_t sa_mask;
    int sa_flags;
    void (*sa_handler)(int);
    void (*sa_sigaction)(int, siginfo_t *, void *);
};
```

where the comments are normative.

It does not make sense to expose the hidden union(s) in a structure specified like this to Rust, because regardless of whether they are real C2011 unnamed unions, fake unnamed unions implemented with macros, or not there at all, they are _undocumented implementation details_ at the API level, and they genuinely do differ from system to system.  The libc crate needs to know what the actual ABI-level implementation is, but nobody else should have to care.

**Comment by carbotaniuman** [comment]
> How does that jive with the claim that unnamed fields are crucially used in some POSIX APIs

I think the answer here is that C11 simply standardized existing practice. Anonymous structs and unions were a common compiler extension and their use in POSIX means C compilers had to support them.

**Comment by RalfJung** [comment]
> It would not make sense for the Rust binding to a structure specified like this to expose the ABI-level hidden union. Regardless of whether the hidden union is a real C2011 unnamed union, a fake unnamed union implemented with macros, or not there at all, it is an undocumented implementation detail at the API level, and it genuinely does differ from system to system. The libc crate needs to know what the actual ABI-level implementation is, but nobody else should have to care.

I don't think it is worth contorting Rust into a shape where these types can be entirely hidden. Even if we accept that Rust has to *somehow* support code of the form `x.sa_handler` for types like this (which I am not entirely convinced of, but it seems like the lang team is), I am personally perfectly fine with telling people to use `__priv` identifiers to hide the extra fields and types hat are used to encode this in Rust.

> I think the answer here is that C11 simply standardized existing practice. Anonymous structs and unions were a common compiler extension and their use in POSIX means C compilers had to support them.

Apparently not, apparently one could just use macros to hack this.

**Comment by retep998** [comment]
My opinion is that this feature is not *necessary* for FFI as replacing unnamed structs/unions with named structs/unions works perfectly fine for simple ABI correctness.

However the very minimal sugar where a field can be marked as unnamed (and thus the subfields can be accessed directly as if they were fields) as proposed by @RalfJung, is sufficiently adequate to fully resolve the quality of life around making FFI in Rust match the original C code. If anything I'd much prefer something this simple over anything more complicated.

**Comment by zackw** [comment]
> Even if we accept that Rust has to somehow support code of the form x.sa_handler for types like this (which I am not entirely convinced of, but it seems like the lang team is)

Do you have an alternative?  On the C side of the fence `x.sa_handler` is the only _portable_ thing one _can_ write; on the Rust side, I don't think anyone wants to write a bunch of `#[cfg]` expressions to know whether they need to write `libc::sigaction { sa_handler = my_handler, ... }` or `libc::sigaction { __sigaction_u = __sigaction_u { sa_handler = my_handler }, ... }` or whatever.

> >  Anonymous structs and unions were a common compiler extension and their use in POSIX means C compilers had to support them.
>
> Apparently not, apparently one could just use macros to hack this.

Both things can be true at the same time :)  I know this _was_ a MSVC extension before it was in C, at least.
